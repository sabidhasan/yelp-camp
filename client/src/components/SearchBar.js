import React from 'react'

class SearchBar extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      // which search result is highlighted, list of search results, whether
      // box has focus, and searchbox contents
      highlightedIndex: -1,
      results: [],
      active: false,
      searchQuery: props.initialValue || ''
    }

    this.doSearch = this.doSearch.bind(this);
    this.setFocus = this.setFocus.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleMouse = this.handleMouse.bind(this);
    this.goToCampground = this.goToCampground.bind(this);
  }

  doSearch(e) {
    // handle arrow key input
    if (e.key === 'ArrowUp') {
      e.preventDefault();
      const newIndex = Math.max(this.state.highlightedIndex - 1, -1);
      this.setState({highlightedIndex: newIndex});
    } else if (e.key === 'ArrowDown') {
      e.preventDefault();
      const newIndex = Math.min(this.state.highlightedIndex + 1, this.state.results.length - 1);
      this.setState({highlightedIndex: newIndex});
    } else if (e.key === 'Enter') {
      // if something is selected then go to it
      this.goToCampground();
    } else {
      // do search
      const icons = {
        'name': '⛺', 'paymentMethods': '💵', 'activities': '🚣',
        'address': '📍', 'description': '📛', 'province': '🌎',
        'region': '🗾'
      }
      this.setState({highlightedIndex: -1});
      fetch(`/search?q=${e.target.value}`)
        .then(res => res.json())
        .then(search => {
          const stateSearch = search.map(v => {
            return {
              id: v.id,
              icon: icons[v.type],
              name: v.campgroundName,
              text: `${v.type}: ${v.excerpt}`}
          })
          // Update the state for results, trimming it to top 5
          this.setState({results: stateSearch.slice(0, 5)})
        })
        .catch(err => console.log("Error occurred while searching ", err))
    }
  }

  handleMouse(key) {
    // Called when mousingover the search results
    this.setState({highlightedIndex: key})
  }

  goToCampground() {
    // Redirects to currently selected campground
    if (this.state.highlightedIndex === -1) {
      window.location = `/search/?q=${this.state.searchQuery}`
    } else {
      window.location = `/campground/${this.state.results[this.state.highlightedIndex].id}`
    }
  }

  handleChange(e) {
    // handles search box controlled component
    this.setState({searchQuery: e.target.value})
  }

  setFocus() {
    // Called when losing/gaining focus to input box
    this.setState({active: !this.state.active});
  }

  render() {
    const searchResults = this.state.results.map((v, i) => (
      <li
        key={i}
        className={i === this.state.highlightedIndex ? 'selected' : undefined}
        onMouseMove={() => this.handleMouse(i)}
        onMouseOut={() => this.setState({highlightedIndex: -1})}
        onMouseDown={this.goToCampground}>
        <i className='search-icon'>{v.icon}</i>
        <div className='searchtext capitalize'>
          <h1>{v.name}</h1><h2>{v.text}</h2>
        </div>
      </li>
    ))
    return (
      <form autoComplete='off' className='searchbar' onSubmit={(e) => e.preventDefault()}>
        <input
          name='search'
          type='text'
          placeholder='Province, Region, City, Campground Name, Amenity, etc.'
          onFocus={this.setFocus}
          onBlur={this.setFocus}
          value={this.state.searchQuery}
          onKeyUp={this.doSearch}
          onChange={this.handleChange}
        />
        <label htmlFor='search'>Find</label>
        <button type='submit' onClick={this.goToCampground}>🔎</button>
        <ul className={this.state.active ? 'results' : 'hidden'}>
          {(searchResults.length || !this.state.searchQuery) ?
            searchResults
            : <p>No results for '{this.state.searchQuery}'</p>
          }
        </ul>
      </form>
    )
  }
}

export default SearchBar
