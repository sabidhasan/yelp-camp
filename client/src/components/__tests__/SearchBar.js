import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import SearchBar from '../SearchBar'

configure({ adapter: new Adapter() });

describe('<SearchBar />', () => {
  let component;
  let props;

  beforeEach(() => {
    const initialQuery = 'Some Search Query'
    props = { initialValue:  initialQuery }
    component = shallow(<SearchBar { ...props } />)
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  });

  it('Initial search query renders in input box', () => {
    expect(component.find('.SearchBar__input').props().value).toBe('Some Search Query');
  })

  it('No results show the no-results message', () => {
    expect(component.find('.SearchBar__no-results')).toHaveLength(1);
  })

  it('Focusing on input box adds active class', () => {
    component.find('.SearchBar__input').prop('onFocus')()
    expect(component.find('.SearchBar__results--active')).toHaveLength(1);
  })

  it('Focusing on input box adds active class', () => {
    component.find('.SearchBar__input').prop('onClick')()
    expect(component.find('.SearchBar__results--active')).toHaveLength(1);
  })

  it('Clicking on input box adds active class', () => {
    component.find('.SearchBar__input').prop('onClick')()
    expect(component.find('.SearchBar__results--active')).toHaveLength(1);
  })

  const getFakeResults = () => {
    const results = [
      {id: 1, icon: 's', name: 'Some Name', text: 'Some search text'},
      {id: 1, icon: 's', name: 'Some Name', text: 'Some search text'}
    ];

    return new Promise((res, rej) => {
      setTimeout(() => {
        res(results);
      }, 500)
    })
  }

  it('Typing in box performs search', async () => {
    const results = await getFakeResults();
    component.setState({ results });
    expect(component.find('.SearchBar__result')).toHaveLength(2);
  })

  it('SearchBar component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
