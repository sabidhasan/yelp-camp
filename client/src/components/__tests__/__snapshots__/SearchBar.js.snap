// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> SearchBar component snapshot matches what is expected 1`] = `
<form
  autoComplete="off"
  className="SearchBar"
  onSubmit={[Function]}
  role="search"
>
  <input
    aria-labelledby="SearchBar__label"
    className="SearchBar__input"
    id="search"
    name="search"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    placeholder="Province, Region, City, Campground Name, Amenity, etc."
    type="text"
    value="Some Search Query"
  />
  <label
    className="SearchBar__label bold flex-center"
    htmlFor="search"
    id="SearchBar__label"
  >
    Find
  </label>
  <button
    aria-label="Search"
    className="SearchBar__button flex-center"
    onClick={[Function]}
    type="submit"
  >
    <i
      className="SearchBar__button-icon fas fa-search"
    />
  </button>
  <ul
    aria-atomic="true"
    aria-hidden="true"
    aria-label="Search results"
    aria-live="assertive"
    className="SearchBar__results--hidden"
    role="listbox"
  >
    <p
      className="SearchBar__no-results bold"
    >
      No results found for '
      Some Search Query
      '
    </p>
  </ul>
</form>
`;
